<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="ns" default=""/>
  <arg name="base_frame"/>
  <arg name="source_pointcloud_topic"/>
  <arg name="target_pointcloud_topic"/>
  <arg name="calibration_service_name" default="extrinsic_calibration"/>

  <arg name="broadcast_calibration_tf" default="false"/>
  <arg name="filter_estimations" default="true"/>

  <arg name="max_calibration_range" default="20.0"/>
  <!--meters -->
  <arg name="max_corr_distance" default="0.5"/>
  <!--meters -->
  <arg name="max_iterations" default="100"/>

  <arg name="min_z" default="0.3"/>
  <!--meters -->
  <arg name="max_z" default="1.2"/>
  <!--meters -->

  <arg name="rviz" default="true"/>
  <let name="rviz_profile" value="$(find-pkg-share lidar_to_lidar_2d_calibrator)/rviz/default.rviz"/>

  <group>
    <push-ros-namespace namespace="$(var ns)"/>
    <node pkg="lidar_to_lidar_2d_calibrator" exec="lidar_to_lidar_2d_calibrator" name="lidar_to_lidar_2d_calibrator" output="screen">
      <remap from="source_input_pointcloud" to="$(var source_pointcloud_topic)"/>
      <remap from="target_input_pointcloud" to="$(var target_pointcloud_topic)"/>
      <remap from="extrinsic_calibration" to="$(var calibration_service_name)"/>

      <param name="base_frame" value="$(var base_frame)"/>

      <param name="broadcast_calibration_tf" value="$(var broadcast_calibration_tf)"/>
      <param name="filter_estimations" value="$(var filter_estimations)"/>

      <param name="max_calibration_range" value="$(var max_calibration_range)"/>
      <param name="max_corr_distance" value="$(var max_corr_distance)"/>
      <param name="max_iterations" value="$(var max_iterations)"/>

      <param name="initial_angle_cov" value="5.0"/>
      <!--degrees -->
      <param name="initial_xy_cov" value="0.05"/>
      <!--meters -->
      <param name="angle_measurement_cov" value="0.5"/>
      <!--degrees -->
      <param name="angle_process_cov" value="0.1"/>
      <!--degrees -->
      <param name="xy_measurement_cov" value="0.005"/>
      <!--meters -->
      <param name="xy_process_cov" value="0.001"/>
      <!--meters -->
      <param name="angle_convergence_threshold" value="0.02"/>
      <!--degrees -->
      <param name="xy_convergence_threshold" value="0.0005"/>
      <!--meters -->

      <param name="min_z" value="$(var min_z)"/>
      <param name="max_z" value="$(var max_z)"/>
    </node>
  </group>

  <node pkg="rviz2" exec="rviz2" name="rviz2" output="screen" args="-d $(var rviz_profile)">
    <remap from="source_pointcloud" to="$(var source_pointcloud_topic)"/>
    <remap from="target_pointcloud" to="$(var target_pointcloud_topic)"/>
  </node>
</launch>
